extensions/data-plane/data-plane-gcs/src/main/java/org/eclipse/dataspaceconnector/gcp/dataplane/gcs/GcsDataSinkFactory.java

```diff
@@ -0,0 +1,107 @@
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.Result;
+import org.eclipse.dataspaceconnector.spi.security.Vault;
+import org.eclipse.dataspaceconnector.spi.types.TypeManager;
+import org.eclipse.dataspaceconnector.spi.types.domain.DataAddress;
+import org.eclipse.dataspaceconnector.spi.types.domain.transfer.DataFlowRequest;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.concurrent.ExecutorService;
+
+public class GcsDataSinkFactory implements DataSinkFactory {
+
+    private final ValidationRule<DataAddress> validation = new GcsSinkDataAddressValidationRule();
+    private final ExecutorService executorService;
+    private final Monitor monitor;
+    private final Vault vault;
+    private final TypeManager typeManager;
+
+
+    public GcsDataSinkFactory(ExecutorService executorService, Monitor monitor, Vault vault, TypeManager typeManager) {
+        this.executorService = executorService;
+        this.monitor = monitor;
+        this.vault = vault;
+        this.typeManager = typeManager;
+    }
+
+    @Override
+    public boolean canHandle(DataFlowRequest request) {
+        return GcsStoreSchema.TYPE.equals(request.getDestinationDataAddress().getType());
+    }
+
+    @Override
+    public @NotNull Result<Boolean> validate(DataFlowRequest request) {
+        var destination = request.getDestinationDataAddress();
+        return validation.apply(destination).map(it -> true);
+    }
+
+    @Override
+    public DataSink createSink(DataFlowRequest request) {
+        var validationResult = validate(request);
+        if (validationResult.failed()) {
+            throw new EdcException(String.join(", ", validationResult.getFailureMessages()));
+        }
+
+        //Get credential from the token if it exists otherwise use the default credentials of the system.
+        var destination = request.getDestinationDataAddress();
+
+        GoogleCredentials googleCredentials;
+
+        if (destination.getKeyName() != null && !destination.getKeyName().isEmpty()) {
+            var credentialsContent = vault.resolveSecret(destination.getKeyName());
+            var gcsAccessToken = typeManager.readValue(credentialsContent, GcpAccessToken.class);
+            googleCredentials = GoogleCredentials.create(
+                    new AccessToken(gcsAccessToken.getToken(),
+                            new Date(gcsAccessToken.getExpiration()))
+            );
+        } else {
+            try {
+                googleCredentials = GoogleCredentials.getApplicationDefault();
+            } catch (IOException e) {
+                throw new GcpException("Error while getting the default credentials.", e);
+            }
+        }
+
+        var storageClient = StorageOptions.newBuilder()
+                .setCredentials(googleCredentials)
+                .build().getService();
```

Comment:
Code from line 77 to 96 is meant to build the `storageClient`, could it be moved into a separated method, this will make the code more readable